### ISSUE TITLE
toy eval ce rejects franken-moe --ckpt-every checkpoints (model vocab -1) — blocks clean CPU held-out CE for the F-series

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#149]
`--ckpt-every` on `train franken-moe` writes `weights/step_N.gguf`, but `toy eval ce --ckpt <that.gguf> --corpus <pack>` rejects it:

```
toy-eval-ce: pack vocab 50257 (TOYC header) != model vocab -1 — wrong pack or wrong model
```

The franken-moe checkpoint GGUF doesn't carry the vocab metadata `eval ce` reads (it sees `model vocab -1`), and `eval ce` expects a "fused-llama checkpoint." So there's no clean way to score a franken-moe checkpoint off-line.

## Why it matters (Tao side)

This is the missing piece that would let held-out CE avoid the in-runner `--eval-corpus` memory spike entirely. Concretely, it just bit F9m: the muon re-run OOM-killed the slice (rc=137) at the end-of-run eval, because in-runner `--eval-corpus` accumulates logits ~`eval-tokens x vocab` (524288 x 50257 ≈ 100 GB host). The whole block-DFA line dodged this with `--ckpt-every` + "separate --load-ckpt eval", but that separate eval turns out not to exist for franken-moe checkpoints — so we fell back to shrinking `--eval-tokens` to 131072 (works, but a smaller eval set than the F9r2 reference).

If `eval ce` accepted franken-moe checkpoints (or `--ckpt-every` wrote the vocab into the GGUF metadata), we'd get: clean CPU held-out CE at full token count, no GPU/unified-memory spike, and byte-stable eval decoupled from the training device. That's the right long-term shape for every held-out-CE read in the F-series.

## Ask

Either (a) have `--ckpt-every` stamp `vocab` (and whatever else `eval ce` needs) into the franken-moe checkpoint GGUF metadata, or (b) teach `toy eval ce` to read a franken-moe checkpoint's vocab. Whichever is cleaner on your side. Not urgent — F9m is running with the reduced-token workaround — but it's the proper fix.

**Comment by OriPekelman** [comment]
Fixed in `0127836`. `make gates` 54/54 green. **You can go back to the full 524288 eval tokens.**

Two things turned out differently from the diagnosis, both in your favour.

## 1. The separate eval already exists

`toy train franken-moe --steps 0 --load-ckpt <ckpt> --eval-corpus <pack> --eval-tokens N` scores a franken-moe checkpoint today, and always did. Verified end to end:

```
$ toy train franken-moe --steps 0 --load-ckpt .../step_4.gguf \
    --eval-corpus data/fineweb_gpt2_smoke.bin --eval-tokens 256 ...
eval_ce: windows=16 tokens=256 ce=10.848929762840271
```

The `--context/--batch/--experts` must match the trained shape (toy#131), which is the only real gotcha.

## 2. The OOM was a per-window LEAK, not `eval-tokens × vocab`

The eval loop called `Toy::Labels.next_token_guarded_batched` **once per window**, and that allocates a fresh `tv × vocab` Mat every call. At your F9m shape (context 256 × batch 32 = 8192 tokens, vocab 50257) that is **3.3 GB of f64 per window** — a 64-window eval asked for ~210 GB. That is the `rc=137`.

It also explains why shrinking `--eval-tokens` *looked* like a fix: fewer windows, less leak. The peak was never driven by the token budget directly.

The training loop immediately above it already keeps **one** buffer and mutates it incrementally — its comment literally calls that "the eval_ce trick" (toy#133), and `eval_ce.rb`'s own header documents the same discipline. The MoE eval lane just never adopted it. It does now.

Peak RSS, context 64 × batch 4:

| windows | before | after |
|---|---|---|
| 4 | 919 MB | 624 MB |
| 16 | 1606 MB | 624 MB |
| 64 | 2784 MB | 624 MB |

Flat in the window count.

**Numeric null:** eval CE is **bit-identical** pre/post at three token counts, checked against the stashed pre-fix binary. Only the buffer management changed, never the label values — so this does not invalidate any eval you have already run, it just makes bigger ones possible.

## What I did NOT do, and why

I did not take option (a) (stamp vocab into the MoE checkpoint). `toy eval ce` builds a `LlamaSeqEngine` and loads **fused-llama tensor names**; a `toy-moe/v1` checkpoint carries expert/router tensors instead. Making the vocab check pass would only move the failure deeper and make it more confusing.

Instead `eval ce` now recognises the architecture and fails loud with the cause and the command that works:

```
toy-eval-ce: <path> is a franken-moe checkpoint (toy-moe/v1), not a fused-llama one —
             `toy eval ce` scores the llama architecture only.
toy-eval-ce: score it with the MoE instrument instead, which reads its own checkpoints:
toy-eval-ce:   toy train franken-moe --steps 0 --load-ckpt <path> --eval-corpus <pack> --eval-tokens N ...
toy-eval-ce: the --context/--batch/--experts must MATCH the trained shape (toy#131).
```

If you still want a true `toy eval ce --arch moe` (one CPU scorer for every lane, decoupled from the training device), say so and I'll file it separately — but with the leak gone, the `--load-ckpt` path already gives you clean CPU held-out CE at full token count, so I don't think it's on the critical path any more.

## Gates

- the rejection must name the architecture **and** point at `train franken-moe --load-ckpt`, and must not say "model vocab -1";
- an RSS-ratio leg pinning that eval memory is flat in the window count — threshold 1.30 against a measured **3.03× before** and **1.00× after**. Linux-guarded (`/usr/bin/time -f` is GNU-only).

**Comment by OriPekelman** [comment]
Closing: shipped in `0127836` — fix(franken-moe): eval allocated a fresh label Mat PER WINDOW — the F9m OOM (toy#149)

Verified by `git log`; the implementing commit names this issue in its subject. Part of the 2026-08-14 issue-state sweep (48 closed, all against a commit in `main`). Reopen if the scope here outran what landed.
